Improved Model.__init__() properties loop.

adamchainz opened the pull request:
* Loop with `.items()` rather than copy the keys into a tuple. This avoids refetching the values (twice).
* Accumulate unrecognized arguments raise an exception with all of them.

kezabelle commented:
So, difficult to get a real flavour of the change, given instantiating a Model always takes me multiple `µs` and the differences presented are not of that magnitude (alas, were it that easy :))

Sketching out notes then, based on picking apart individual bits (hopefully correctly!):

```
x = {'1': 2, 'test': 3, 'whee': 4, 'another': 5, '6': 7, '8': 9}
y = {'1': 2}
```

- When using `y`, such that `len(kwargs) == 1` 
    - both `for prop in tuple(kwargs): pass` and `for prop, value in kwargs.items(): pass` take roughly the same time for me, roughly `130ns`
- When using `x`, making `len(kwargs) == 6`:
    - the tuple form comes in at `190ns`
    - while the items version ends up at `210ns`.

Given the same dictionaries above, and using:
```
for prop in tuple(x|y):
    x|y[prop]
    x|y[prop]
```

- Iterating and accessing `kwargs[prop]` _twice_ for a dictionary of **six** elements costs `502ns` (or `345ns` for _one_ access).
- Iterating and accessing `kwargs[prop]` _twice_ for a dictionary of **one** element though, costs only `190ns` (or `164ns` for _one_ access)
- The cost is getting the value is naturally "free" in the `.items()` form, because it's unrolled as part of the loop.

I _think_ once you take into account the actual accessing of the variable (which can actually be _three_ times, non? `if` `setattr` and `del`), for any situation where `len(kwargs) > 1`, the items version & getting the value for "free" is faster. And I _do_ like the multiple errors at once better for user ergonomics...

Here's my timings for various `kwargs` lengths, using the following timeit snippets, which I think are hopefully representative, but may not be perfect:

```
for prop in tuple(obj):
    obj[prop]
    obj[prop]

for prop, value in obj.items():
    pass
```
| length 	| tuple 	| items 	|
|---		|---		|---		|
| 1 		| 187ns 	| 133ns 	|
| 2 		| 258ns 	| 149ns 	|
| 3 		| 331ns 	| 173ns 	|
| 4 		| 419ns 	| 194ns 	|
| 5 		| 465ns 	| 199ns 	|
| 6 		| 567ns 	| 221ns 	|

----

But to complicate things further, accumulating `prop` into a new list (`unexpected`) looks to be slightly slower than mutating and deleting keys (`del kwargs[prop]`) from the original dictionary (`kwargs`).

Using the below (which naively assumes _all_ are unexpected arguments):
```
a = {...}
b = []
for k, v in a.items():
    b.append(k)

a = {...}
for _ in tuple(a):
    del a[_]
```

I get these kinds of results:

| length 	| del 	| append 	|
|---		|---		|---		|
| 1 		| 206ns 	| 245ns 	|
| 3 		| 321ns 	| 399ns 	|
| 6 		| 481ns 	| 669ns 	|

So it may be a case of what _wants_ to be optimised for. If its the happy path of _has remaining kwargs and they correspond to properties or virtual fields_, I want to say this probably makes sense (and it's not the _common_ case anyway, because it's _remaining_ kwargs, right?). If its the unhappy path of _user has passed multiple nonsense kwargs_, I **think** it's slightly slower but the ergonomics of being told up-front all the invalid arguments you've provided is possibly still nicer? Certainly I hate it when I make a change and there's _still_ an exception waiting to raise ;)

Please, feel free to point out any errors in my thinking or profiling - I'll probably learn something. All data was calculated using IPython's `%%timeit` magic, for better or worse.

carltongibson reviewed (commented):
> ...given instantiating a Model always takes me multiple µs and the differences presented are not of that magnitude...

Just assessing on readability, I think I prefer the suggested version.

felixxm commented on django/db/models/base.py:
```diff
@@ -513,18 +513,21 @@ def __init__(self, *args, **kwargs):
-                    pass
-            for kwarg in kwargs:
-                raise TypeError("%s() got an unexpected keyword argument '%s'" % (cls.__name__, kwarg))
+                    unexpected.append(prop)
```
Do we need to catch `AttributeError`? It looks unnecessary since d2a26c1a90e837777dabdf3d67ceec4d2a70fb86 :thinking:

adamchainz replied:
I believe you're right - it could only occur from properties, which are already handled.

adamchainz replied:
Dropping `AttributeError` causes `properties.tests.PropertyTests.test_setter` to fail, because it expects the `TypeError` instead of `AttributeError`. IMO allowing the `AttributeError` to be raised is fine, it highlights that the property is the problem. Changing the test.

adamchainz commented:
> (alas, were it that easy :))

Yes if only!

> But to complicate things further, accumulating `prop` into a new list (`unexpected`) looks to be slightly slower than mutating and deleting keys (`del kwargs[prop]`) from the original dictionary (`kwargs`).
> Note that being slower in the unhappy path is basically fine in and of itself (I think), because it'll only ever fire once unless you're choosing to silence the exception it throws.

💯

The happy path, which 99.999% of model instantiations will go through, is the one to minimize. Eliminating key deletion there is a win, even if we add work for the unhappy path.

I've just changed the PR to make `unexpected` a `tuple`. This is basically free to create (actually cached at compile time in the function consts) but even slower in the unhappy path.

> Just assessing on readability, I think I prefer the suggested version.

I think I managed to make it less readable with my recent push 😅. I added two changes:

* Make `unexpected` a tuple. 
* [Limited `try/except`](http://adamj.eu/tech/2019/10/02/limit-your-try-clauses-in-python/) to just the `get_field()` call, which is the only thing that can raise `FieldDoesNotExst`. This is a bit less readable but it will also prevent swallowing any bugs that might occur from e.g. fields whose `__setattr__` ends up touching other fields that don't exist. Happy to revert to the `try` around the `if` if you think it's not worth it.

felixxm reviewed (approved):
@adamchainz Thanks :+1:

ngnpope commented:
> > (alas, were it that easy :))
>
> Yes if only!

@kezabelle I notice that you often mention the instability in results making it hard to tell whether there is actually an improvement. Perhaps look into the following:

- https://pyperformance.readthedocs.io/usage.html#how-to-get-stable-benchmarks
- https://pyperf.readthedocs.io/en/latest/run_benchmark.html#how-to-get-reproducible-benchmark-results
- https://pyperf.readthedocs.io/en/latest/system.html#system

I've had pretty stable results by just using `python -m pyperf system tune`. You can use this and then decide not to use `pyperf` and work with `timeit`. It's just a handy way to get things tuned. Switch `tune` for `reset` to undo the changes.

kezabelle commented:
Hey @ngnpope,
This is the wrong avenue to continue the conversation, but just to let you know, I didn't ignore this, but it wasn't immediately/obviously useful regrettably (see below)... I'll have to investigate further at some point. 

> [...] I've had pretty stable results by just using `python -m pyperf system tune`. You can use this and then decide not to use `pyperf` and work with `timeit`. It's just a handy way to get things tuned. Switch `tune` for `reset` to undo the changes.

All I get is:
```
> python3 -m pyperf system tune
WARNING: no operation available for your platform
```
Computers are hard :)
